File: Data-Structure/Java/LeetCode/Array/844. Backspace String Compare.java
/*
Given two strings s and t, return true if they are equal when both are typed into empty text editors. '#' means a backspace character.

Note that after backspacing an empty text, the text will continue empty.

Example 1: Input: s = "ab#c", t = "ad#c" Output: true
Explanation: Both s and t become "ac".
Example 2: Input: s = "ab##", t = "c#d#" Output: true
Explanation: Both s and t become "".
Example 3: Input: s = "a##c", t = "#a#c" Output: true
Explanation: Both s and t become "c".
Example 4: Input: s = "a#c", t = "b" Output: false
Explanation: s becomes "c" while t becomes "b".
*/
class Solution {
    public boolean backspaceCompare(String S, String T) {
        int i = S.length() - 1, j = T.length() - 1;
        int skipS = 0, skipT = 0;
        while (i >= 0 || j >= 0) {
            while (i >= 0) { // Find position of next possible char in build(S)
                if (S.charAt(i) == '#') {
                    skipS++;
                    i--;
                } else if (skipS > 0) {
                    skipS--;
                    i--;
                } else
                    break;
            }
            while (j >= 0) { // Find position of next possible char in build(T)
                if (T.charAt(j) == '#') {
                    skipT++;
                    j--;
                } else if (skipT > 0) {
                    skipT--;
                    j--;
                } else
                    break;
            }
            // If two actual characters are different
            if (i >= 0 && j >= 0 && S.charAt(i) != T.charAt(j)) {
                return false;
            }
            // If expecting to compare char vs nothing
            if ((i >= 0) != (j >= 0)) {
                return false;
            }
            i--;
            j--;
        }
        return true;
    }
}
//stack
class Solution {
    public boolean backspaceCompare(String S, String T) {
        return build(S).equals(build(T));
    }

    public String build(String S) {
        Stack<Character> ans = new Stack();
        for (char c : S.toCharArray()) {
            if (c != '#')
                ans.push(c);
            else if (!ans.empty())
                ans.pop();
        }
        return String.valueOf(ans);
    }
}

File: Data-Structure/_Algorithm (LeetCode)/Alg-Study-Plan_Algorithm II/Day 4 Two Pointers/00844. Backspace String Compare/844. Backspace String Compare.js
// 844. Backspace String Compare
/*
Given two strings s and t, return true if they are equal when both are typed into empty text editors. '#' means a backspace character.

Note that after backspacing an empty text, the text will continue empty.

 

Example 1:

Input: s = "ab#c", t = "ad#c"
Output: true
Explanation: Both s and t become "ac".
Example 2:

Input: s = "ab##", t = "c#d#"
Output: true
Explanation: Both s and t become "".
Example 3:

Input: s = "a#c", t = "b"
Output: false
Explanation: s becomes "c" while t becomes "b".
*/
{
    // T: O( + N), S: O(M + N)
    var backspaceCompare = function (S, T) {
        const backspaceHelper = (str) => {
            let stock = [];
            for (let char of str) {
                (char === '#') ? stock.pop() : stock.push(char);
            }
            return stock.join("");
        }

        return backspaceHelper(S) === backspaceHelper(T);
    };
    //Crawler Log Folder

}